pkg/containerruntimediscovery/containerd/discoverer.go

```diff
@@ -0,0 +1,516 @@
+		client:          client,
+		snapshotterName: DefaultSnapshotterName,
+		namespace:       DefaultNamespace,
+		skipUnpackImage: DefaultSkipUnpackImage,
+	}, nil
+}
+
+type onFound func(image containerd.Image) (bool, error)
+
+func (d *discoverer) imageIDWalk(ctx context.Context, imageID string, f onFound) (bool, error) {
+	var found bool
+
+	// ContainerD doesn't allow to filter images by config digest, so we
+	// have to walk all the images to find all the images by ID and then
+	// merge them together.
+	images, err := d.client.ListImages(ctx)
+	if err != nil {
+		return found, fmt.Errorf("failed to list images: %w", err)
+	}
+
+	for _, image := range images {
+		configDescriptor, err := image.Config(ctx)
+		if err != nil {
+			return found, fmt.Errorf("failed to load image config descriptor: %w", err)
+		}
+		id := configDescriptor.Digest.String()
+
+		if id != imageID {
+			continue
+		}
+
+		found = true
+
+		stop, err := f(image)
+		if err != nil {
+			return found, err
+		}
+
+		if stop {
+			break
+		}
+	}
+
+	return found, nil
+}
+
+func (d *discoverer) Images(ctx context.Context) ([]models.ContainerImageInfo, error) {
+	logger := log.GetLoggerFromContextOrDefault(ctx)
+
+	images, err := d.client.ListImages(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list images: %w", err)
+	}
+
+	imageSet := map[string]models.ContainerImageInfo{}
+	for _, image := range images {
+		// Ignore our transient images used for snapshoting the
+		// containers they will be cleaned up as soon as the export is
+		// done.
+		if strings.HasPrefix(image.Name(), "vmclarity.io/container-snapshot:") {
+			continue
+		}
+
+		imageInfo, err := d.getContainerImageInfo(ctx, image)
+		if err != nil {
+			return nil, fmt.Errorf("unable to convert image %s to container image info: %w", image.Name(), err)
+		}
+
+		if imageInfo.ImageID == "" {
+			logger.Warnf("found image with empty ImageID: %s", imageInfo.String())
+			continue
+		}
+
+		existing, ok := imageSet[imageInfo.ImageID]
+		if ok {
+			merged, err := existing.Merge(imageInfo)
+			if err != nil {
+				return nil, fmt.Errorf("unable to merge image %v with %v: %w", existing, imageInfo, err)
+			}
+			imageInfo = merged
+		}
+		imageSet[imageInfo.ImageID] = imageInfo
+	}
+
+	result := []models.ContainerImageInfo{}
+	for _, image := range imageSet {
+		result = append(result, image)
+	}
+
+	return result, nil
+}
+
+func (d *discoverer) Image(ctx context.Context, imageID string) (models.ContainerImageInfo, error) {
+	var result models.ContainerImageInfo
+	found, err := d.imageIDWalk(ctx, imageID, func(image containerd.Image) (bool, error) {
+		cii, err := d.getContainerImageInfo(ctx, image)
+		if err != nil {
+			return false, fmt.Errorf("unable to convert image %s to container image info: %w", image.Name(), err)
+		}
+
+		result, err = result.Merge(cii)
+		if err != nil {
+			return false, fmt.Errorf("unable to merge image %v with %v: %w", result, cii, err)
+		}
+
+		return false, nil
```

Comment:
Actually the logic implemented here is correct, it just really confusing. So this walk function should never return with `true` stopping the walk as we want to collect all the images with the same ID and merge their information together into a single `ContainerImageInfo` object which can represent a single image with multiple different repo tags. In case of `containerd` we need to iterate over all the repo tags to find the ones which belong to the same image ID.

I'll refactor this a bit to make it more readable.